### DISCUSSION TITLE
Testing of external MooseMesh?

### DISCUSSION THREAD

**Discussion by aprilnovak** [idaholab/moose#16181] [Q&A Tools] [answered]
Hi all,

I'm wondering if there are any built-in classes for testing meshes? I'm wrapping a CFD code through MOOSE's ExternalProblem interface, and I want to write some unit tests to confirm that our NekMesh class works properly.

The doc for MooseObjectUnitTest seems to suggest that mesh-based objects like kernels can't be tested this way - does that also apply to the mesh objects themselves? I'm also considering the regression test route, by perhaps creating a new `test/src/postprocessor/nElems.C`, etc. for all the mesh data I want to check, but I thought the unit test approach could be easier.

Thanks,

-April

**Comment by aeslaughter** [comment, marked as answer]
Making the unit test work for this will be difficult. What I have done is just create a test UserObject. Register it to your test application. Then you can do whatever crazy checks you want. I often make a MooseEnum that lists out the possible tests and then in constructor or initialSetup have a massive if statement that will check each item. Finally, create an input file and then add a bunch of blocks in your Tests file that hits everything you need.

**Reply by aprilnovak** [reply]
Thanks, that's a good idea - hadn't thought of that! This should be perfect for my needs.

**Reply by bielsnohr** [reply]
Apologies for opening an old resolved discussion, but this sounds ideal for facilitating something closer to TDD in my app. Could either of you @aprilnovak @aeslaughter  point to an example of this testing strategy in practice?

**Reply by aprilnovak** [reply]
Hi @bielsnohr, we currently use this UserObject approach in Cardinal, a wrapping of nekRS as a MOOSE application. The repository is private, so I can't point you to a complete example at the moment. But I can try to describe the general approach.

We made a `NekMeshInfoPostprocessor` that inherits from `GeneralPostprocessor` that we have a `MooseEnum` to toggle between various things we want to test:

```
MooseEnum test_type("num_elems num_nodes node_x node_y node_z");  
params.addRequiredParam<MooseEnum>("test_type", test_type, "The type of evaluation");
```

And then in `getValue()`, we just have a large if statement that returns the value we're interested in depending on the enumeration. By downcasting the mesh to our external mesh class we're testing (`NekRSMesh`), then we can call any of the methods on that external mesh for testing that it's set up correctly.

```
_nek_mesh = dynamic_cast<const NekRSMesh *>(&_mesh);

if (_test_type == "num_elems")
  return _nek_mesh->nElem();
else if (_test_type == "num_nodes")
  return _nek_mesh->nNodes();
// ...
```

Then, in a test, simply add the postprocessors you want to test - save the output file as a gold file, and then those postprocessors will be tested.

```
[Postprocessors]
  [./num_elems]
    type = NekMeshInfoPostprocessor
    test_type = num_elems
  [../]
  [./num_nodes]
    type = NekMeshInfoPostprocessor
    test_type = num_nodes
  [../]
[]
```

**Reply by bielsnohr** [reply]
Brilliant, thank you @aprilnovak for the detailed information. Quick question on implementation: is `NekMeshInfoPostprocessor` source implemented in the `test/` directory, or in the usual location with the rest of the `MooseApp`?

**Reply by aprilnovak** [reply]
If you will only use your postprocessor for testing, it makes sense to put it in the `test/src/postprocessors` directory to help differentiate its purpose from other postprocessors in your application. But the choice is up to you, you can also place the postprocessor in `src/postprocessors` with the other source files of your app

**Reply by loganharbour** [reply]
I'd also like to note that if you're looking at information like April posted up above, it's worth looking at a test Receiver as well. This'll let you output a ton of mixed type data/maps/etc and run simple diffs on it. We do this in MeshInfo.

**Reply by bielsnohr** [reply]
@loganharbour looking at the MOOSE Doxygen, the only stated purpose of the Receiver is to modify the value coming from at PostProcessor. I don't really have a use case for why that is useful. Do you have an example for MeshInfo? A quick search of GitHub suggests this project might be closed source.

**Reply by loganharbour** [reply]
Gah... I was working on a Receiver object at the time and I typed that poorly. Bad multi tasking! What I should have typed instead is:

> I'd also like to note that if you're looking at information like April posted up above, it's worth looking at a test **Reporter** as well. This'll let you output a ton of mixed type data/maps/etc and run simple diffs on it. We do this in MeshInfo.

MeshInfo is here: https://github.com/idaholab/moose/blob/next/framework/src/reporters/MeshInfo.C

Reporters are a little more complex as they can support arbitrary output in JSON format (which is the primary purpose). Example output is here: https://github.com/idaholab/moose/blob/next/test/tests/reporters/mesh_info/gold/mesh_info_out.json

The Reporter is quite a bit more complex because it doesn't just involve scalar values, but it is much more versatile as a result of this. As a whole, we would like to transition more complex output to the Reporter system as a result of this versatility. If you give us an idea of what type of data you would like to output for test purposes, we should be able to pretty quickly tell you how easy of an implementation it would be.

**Reply by bielsnohr** [reply]
@loganharbour  not to worry, I'm all to familiar with humans being deficient at concurrency, myself included :wink: 

Thanks for the offer of advice on my particular use case. Brief summary is that I will be reading scalar values on the elements of a trig mesh from an HDF5 file. The scalar values are held in a 1D array. I want to proceed in a stepwise fashion somewhat like TDD:

1. write test to check array has been read correctly from HDF5, then implement read of array
2. write test to check array has been mapped correctly into the AuxVariable of the AuxiliarySystem, then implement copy from array to AuxVariable
3. write test to check integrated value of the AuxVariable

All of this will be happening within the `syncSolutions()` method of an ExternalProblem subclass. Hope that is clear.

**Reply by aeslaughter** [reply]
I also think that is appropriate to make a test object that just hard codes the "gold" information in the C++. If the data doesn't match throw an error. I even add sucess error messages like "read of HDF5 passes" that get thrown if it all worked as expected and setup the test specification to expect that "error"
